src/pal/src/misc/cgroup.cpp

```diff
@@ -0,0 +1,275 @@
+
+/*++
+
+Module Name:
+
+    cgroup.cpp
+
+Abstract:
+    Read memory limits for the current process
+--*/
+
+#include "pal/dbgmsg.h"
+SET_DEFAULT_DEBUG_CHANNEL(MISC);
+#include "pal/palinternal.h"
+#include <sys/resource.h>
+
+#define MAX_SIZE_T (~(SIZE_T)0)
+#define PROC_MOUNTINFO_FILENAME "/proc/self/mountinfo"
+#define PROC_CGROUP_FILENAME "/proc/self/cgroup"
+#define PROC_STATM_FILENAME "/proc/self/statm"
+#define MEM_LIMIT_FILENAME "/memory.limit_in_bytes"
+#define MEM_SWAP_LIMIT_FILENAME "/memory.memsw.limit_in_bytes"
+
+class CGroup
+{
+    char _memory_cgroup_path[MAX_PATH];
+public:
+    CGroup()
+    {
+        if(!findMemoryHierarchyMount(_memory_cgroup_path, MAX_PATH))
+        {
+            _memory_cgroup_path[0] = '\0';
+            return;
+        }
+        char cgroup_path_relative_to_mount[MAX_PATH];
+        if(!findCGroupPathForMemorySubsystem(cgroup_path_relative_to_mount, MAX_PATH))
+        {
+            _memory_cgroup_path[0] = '\0';
+            return;
+        }
+        strcat_s(_memory_cgroup_path, MAX_PATH, cgroup_path_relative_to_mount);
+    }
+    
+    CGroup(const char* memory_cgroup_path)
+    {
+        strcpy(_memory_cgroup_path, memory_cgroup_path);
+    }
+
+    const char* path()
+    {
+        return _memory_cgroup_path;
+    }
+
+    bool physicalMemoryLimit(size_t &val)
+    {
+        char mem_limit_filename[MAX_PATH];
+
+        if(_memory_cgroup_path[0] == 0)
+            return false;
+
+        strcpy_s(mem_limit_filename, MAX_PATH, _memory_cgroup_path);
+        strcat_s(mem_limit_filename, MAX_PATH, MEM_LIMIT_FILENAME);
+        return readMemoryValueFromFile(mem_limit_filename, val);
+    }
+    bool swapMemoryLimit(size_t &val)
+    {
+        char memsw_limit_filename[MAX_PATH];
+
+        if(_memory_cgroup_path[0] == 0)
+            return false;
+
+        strcpy_s(memsw_limit_filename, MAX_PATH, _memory_cgroup_path);
+        strcat_s(memsw_limit_filename, MAX_PATH, MEM_SWAP_LIMIT_FILENAME);
+        return readMemoryValueFromFile(memsw_limit_filename, val);
+    }
+private:
+    bool findMemoryHierarchyMount(char* mountpath, size_t mountpathsize)
+    {
+        bool retval = false;
+        char *line = nullptr;
+        size_t lineLen = 0;
+        FILE *mountinfofile = fopen(PROC_MOUNTINFO_FILENAME, "r");
+        if(mountinfofile == NULL)
+            goto done;
+
+        while(!retval && getline(&line, &lineLen, mountinfofile) != -1)
+        {
+            char filesystemType[100];
+            char options[MAX_PATH];
+
+            char* separatorChar = strchr(line, '-');
+
+            // See man page of proc to get format for /proc/self/mountinfo file
+            int sscanfRet = sscanf_s(separatorChar, 
+                                     "- %s %*s %s",
+                                     filesystemType, (unsigned)_countof(filesystemType),
+                                     options, (unsigned)_countof(options));
+            if (sscanfRet != 2)
```

Comment:
Could you please unify the spacing style between the if / while and the following `(`? Sometimes you have space there and sometimes not. I prefer the style with space, but I'll leave it up to you.